### PR TITLE
Create ast.variable and bug fixes

### LINKED ISSUE #490: pyccel/ast/core.py is larger than codacy's maximum file size
pyccel/ast/core.py is larger than codacy's maximum file size (150KB as specified here https://docs.codacy.com/faq/troubleshooting/why-is-my-file-missing/#file-size-limitation-on-codacy-cloud). It is currently 152kB.

We should either find a way to break it into pieces. Eg. we could create an arithmetic.py file for all basic arithmetic types

Or we should use the Codacy Analysis CLI locally

### PR DESCRIPTION
**Pull request by EmilyBourne**
Move classes necessary for the handling of Variables into a new file `ast.variable`
Fixes #490 (ast/core.py too large for codacy).

**Commit Summary**
- Move classes necessary for the handling of Variables into a new file `ast.variable`
- Move Slice to `ast.internals` (should this be in operators?)
- Remove `is_polymorphic` property from Variables. This property exists in the class dtype and a "polymorphic variable" doesn't make much sense
- Add `passed_from_dotted` property to `Declare`. In fortran types are declared differently when they are used for arguments which are passed due to class syntax
- General Codacy fixes

### CODE REVIEW

**Comment by EmilyBourne** [comment]
elemental functions can be tested for c if a print function is written for `nd_array` type

File: pyccel/ast/variable.py [hunk 1]
```diff
@@ -0,0 +1,848 @@
...
+
+    @is_target.setter
+    def is_target(self, is_target):
+        if not isinstance(is_target, bool):
+            raise TypeError('is_target must be a boolean.')
+        self._is_target = is_target
+
+    @property
+    def is_polymorphic(self):
+        return self._is_polymorphic
```
Comment (EmilyBourne) [inline pyccel/ast/variable.py:313]: @saidctb this needs a docstring. Can you provide more information? I am not sure what a polymorphic variable is. This seems like something that should be the property of the dtype?

Comment (saidctb) [reply pyccel/ast/variable.py:313]: it's used for classes,  but I'm not sure if it's used at all

Comment (EmilyBourne) [reply pyccel/ast/variable.py:313]: It is used once in the fortran printer but it seems to also exist in the dtype. I think I will leave it just in the dtype if that is ok with you

Comment (saidctb) [reply pyccel/ast/variable.py]: yes fine

File: pyccel/ast/variable.py [hunk 2]
```diff
@@ -0,0 +1,830 @@
...
+    True
+    """
+
+    def __new__(
+        cls,
+        base,
+        *args,
+        **kw_args
+        ):
+
```
Comment (MenoIy) [inline pyccel/ast/variable.py:716]: there was a check for base argument here it is no more needed?

**Review by MenoIy — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/fcode.py [hunk 3]
```diff
@@ -1564,13 +1559,15 @@ def function_signature(self, expr, name):
...
-                elif str(arg) == 'self':
+                if i == 0 and expr.cls_name:
+                    dec = Declare(arg.dtype, arg, intent='inout', passed_from_dotted = True)
+                elif expr.arguments_inout[i]:
                     dec = Declare(arg.dtype, arg, intent='inout')
                 else:
                     dec = Declare(arg.dtype, arg, intent='in')
                 args_decs[str(arg)] = dec
+        print(args_decs)
+        print([d.passed_from_dotted for d in args_decs.values()])
```
Comment (MenoIy) [inline pyccel/codegen/printing/fcode.py:1570]: trailing print

File: pyccel/codegen/printing/fcode.py [hunk 4]
```diff
@@ -1636,6 +1633,8 @@ def _print_FunctionDef(self, expr):
         for v in vars_to_print:
             if (v not in expr.local_vars) and (v not in expr.results) and (v not in expr.arguments):
                 decs[str(v)] = Declare(v.dtype,v)
+        print("decs : ",decs)
+        print(expr.arguments)
```
Comment (MenoIy) [inline pyccel/codegen/printing/fcode.py:1637]: here too

File: pyccel/codegen/printing/fcode.py [hunk 5]
```diff
@@ -559,6 +558,7 @@ def _print_DottedVariable(self, expr):
             return self._print(var) + '%' +self._print(expr.name)
         else:
             return self._print(expr.lhs) + '%' +self._print(expr.name)
+        return self._print(expr.lhs) + '%' +self._print(expr.name)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:561]: No it's just wrong. I suspect github merged it automatically given that the changes to DottedVariable were originally done here

File: pyccel/codegen/printing/fcode.py (hunk 4 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:1637]: oops

File: pyccel/ast/variable.py (hunk 2 above)
Comment (EmilyBourne) [reply pyccel/ast/variable.py:716]: The check was quite large and referred to lots of sympy types that should only exist in the syntactic stage. But IndexedElement is only used in the semantic stage so I don't see the relevance. The last PR also added a check for the semantic stage to test that base is a variable. But this condition is not strictly necessary. E.g:
```
a = f()[0]
```
We may wish to forbid such syntax, but in that case a neat error should be raised

**Review by MenoIy — APPROVED** [review]

**Review by saidctb — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Comment by yguclu** [comment]
Some conflicts need to be resolved before proceeding with the merge.

**Review by yguclu — APPROVED** [review]

**Pull request merged**